arch/risc-v/src/bl808/bl808_wdt.c

```diff
@@ -0,0 +1,385 @@
+   */
+
+  modifyreg32(BL808_TIMER_TCCR(priv->idx), WDT_CLKSEL_MASK,
+              WDT_CLK_SRC_NONE << WDT_CLKSEL_SHIFT);
+
+  priv->started = false;
+  return OK;
+}
+
+int bl808_wdt_keepalive(FAR struct watchdog_lowerhalf_s *lower)
+{
+  struct bl808_wdt_s *priv = (struct bl808_wdt_s *)lower;
+
+  /* Check that the watchdog is running */
+
+  if (priv->started == false)
+    {
+      return -EIO;
+    }
+
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COUNT_CLEAR(priv->idx), 0, WDT_CLEAR_COUNT);
+  while (getreg32(BL808_WDT_COUNTER(priv->idx)) != 0);
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COUNT_CLEAR(priv->idx), WDT_CLEAR_COUNT, 0);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: bl808_wdt_getstatus
+ *
+ * Description:
+ *   Get current watchdog status. Returns to status parameter.
+ *
+ ****************************************************************************/
+
+int bl808_wdt_getstatus(FAR struct watchdog_lowerhalf_s *lower,
+                          FAR struct watchdog_status_s *status)
+{
+  struct bl808_wdt_s *priv = (struct bl808_wdt_s *)lower;
+
+  status->flags = priv->started
+    | ((priv->callback == NULL) << 1)
+    | ((priv->callback != NULL) << 2);
+
+  status->timeout = priv->timeout;
+
+  uint32_t current_count = getreg32(BL808_WDT_COUNTER(priv->idx));
+  status->timeleft = priv->timeout - current_count;
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: bl808_wdt_settimeout
+ *
+ * Description:
+ *   Set a new timeout value and reset the watchdog.
+ *
+ ****************************************************************************/
+
+int bl808_wdt_settimeout(FAR struct watchdog_lowerhalf_s *lower,
+                           uint32_t timeout)
+{
+  struct bl808_wdt_s *priv = (struct bl808_wdt_s *)lower;
+
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COMP(priv->idx), 0xffff, timeout);
+  priv->timeout = timeout;
+
+  /* Clock is needed to clear counter (assuming same behavior as GP timers) */
+
+  modifyreg32(BL808_TIMER_TCCR(priv->idx), WDT_CLKSEL_MASK,
+              WDT_CLK_SRC_1K << WDT_CLKSEL_SHIFT);
+
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COUNT_CLEAR(priv->idx), 0, WDT_CLEAR_COUNT);
+  while (getreg32(BL808_WDT_COUNTER(priv->idx)) != 0);
+
+  /* Disable clock to stop timer from running after clear */
+
+  modifyreg32(BL808_TIMER_TCCR(priv->idx), WDT_CLKSEL_MASK,
+              WDT_CLK_SRC_NONE << WDT_CLKSEL_SHIFT);
+
+  BL808_UNLOCK_WDT(priv->idx);
+  modifyreg32(BL808_WDT_COUNT_CLEAR(priv->idx), WDT_CLEAR_COUNT, 0);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: bl808_wdt_capture
+ *
+ * Description:
+ *   Assigns a callback to be called on timeout. If
+ *   the callback is null, configure the watchdog
+ *   as a reset source.
+ *
```

Comment:
Ditto

Reply:
I wasn't sure what capture was supposed to return from watchdog.h, but I found out from looking at other drivers. I'll add the proper return and edit the function description.